### PR TITLE
Factor Windows PDB CDI encoding to a separate shareable type

### PR DESCRIPTION
**Pull request by tmat**

### CODE REVIEW

**Comment by tmat** [comment]
@dotnet/roslyn-compiler Please review.

File: src/Compilers/Core/CodeAnalysisTest/Emit/CustomDebugInfoTests.cs [hunk 1]
```diff
@@ -369,6 +371,222 @@ public void EncCdiAlignment()
...
+                // aligned record size
+                0x10, 0x00, 0x00, 0x00,
+                
+                // payload (4B aligned)
+                0x03, 0x00, // bucket count
+                0x01, 0x00, // using count #1
+                0x02, 0x00, // using count #2
+                0x03, 0x00, // using count #3
+            }, cdi);
+        }
```
Comment (cston) [inline src/Compilers/Core/CodeAnalysisTest/Emit/CustomDebugInfoTests.cs]: Is this test different from `UsingInfo2`?

File: src/Compilers/Core/CodeAnalysisTest/Emit/CustomDebugInfoTests.cs [hunk 2]
```diff
@@ -369,6 +371,222 @@ public void EncCdiAlignment()
...
+
+                // aligned record size
+                0x14, 0x00, 0x00, 0x00,
+                
+                // payload (4B aligned)
+                0x04, 0x00, // bucket count
+                0x01, 0x00, // using count #1
+                0x02, 0x00, // using count #2
+                0x03, 0x00, // using count #3
+                0x04, 0x00, // using count #3
```
Comment (cston) [inline src/Compilers/Core/CodeAnalysisTest/Emit/CustomDebugInfoTests.cs]: `// using count #4`

**Comment by cston** [comment]
LGTM

File: src/Compilers/Core/CodeAnalysisTest/Emit/CustomDebugInfoTests.cs (hunk 1 above)
Comment (tmat) [inline src/Compilers/Core/CodeAnalysisTest/Emit/CustomDebugInfoTests.cs]: Yes, this one has just 3 groups and no alignment. The previous has 4 groups and alignment.

File: src/Compilers/Core/CodeAnalysisTest/Emit/CustomDebugInfoTests.cs (hunk 2 above)
Comment (tmat) [inline src/Compilers/Core/CodeAnalysisTest/Emit/CustomDebugInfoTests.cs]: Will fix.

**Comment by tmat** [comment]
@dotnet/roslyn-compiler Ping

File: src/Dependencies/CodeAnalysis.Metadata/CustomDebugInfoKind.cs [hunk 3]
```diff
@@ -7,11 +7,11 @@
     /// </summary>
     internal enum CustomDebugInfoKind : byte
     {
-        UsingInfo = 0,
-        ForwardInfo = 1,
-        ForwardToModuleInfo = 2,
+        UsingGroups = 0,
```
Comment (AlekseyTs) [inline src/Dependencies/CodeAnalysis.Metadata/CustomDebugInfoKind.cs]: >UsingGroups [](start = 8, length = 11)

Please add comments for each member.

**Comment by AlekseyTs** [comment]
LGTM

File: src/Dependencies/CodeAnalysis.Metadata/CustomDebugInfoKind.cs (hunk 3 above)
Comment (tmat) [inline src/Dependencies/CodeAnalysis.Metadata/CustomDebugInfoKind.cs]: Will do.

**Comment by tmat** [comment]
Test windows_release_vs-integration_prtest  please
https://github.com/dotnet/roslyn/issues/17631

**Comment by tmat** [comment]
test windows_release_vs-integration_prtest please

**Comment by tmat** [comment]
test windows_release_vs-integration_prtest please

**Comment by tmat** [comment]
test windows_debug_vs-integration_prtest please

**Pull request merged**